src/sinks/datadog/traces/sink.rs

```diff
@@ -113,7 +120,33 @@ where
             })
             .into_driver(self.service);
 
-        sink.run().await
+        sink.run().await?;
+
+        // Signals the APM stats flushing thread to send the remaining cached payloads and exit.
+        {
+            // create a channel for the stats flushing thread to communicate back that it has flushed
+            // remaining stats. This is necessary so that we do not terminate the process while the
+            // stats flushing thread is trying to complete the HTTP request.
+            let (sender, mut receiver) = channel();
+
+            // Notify the stats thread task to flush remaining payloads and shutdown.
+            let _ = self.shutdown.send(sender);
+
+            // allow the stats flushing thread up to 50 seconds to send the remaining stats payloads
+            // Vector allows 59 seconds for components to shutdown. We shouldn't need more than 10 but
+            // doesn't hurt to be conservative for the timeout since we are using the oneshot.
+            let timeout = sleep(Duration::from_secs(50));
```

Comment:
Vector is already going to wait up to 60 seconds for all components to shut down, so if we're having it sleep just nearly as long, we might as well just let it sleep as long as possible -- the global shutdown grace period -- by not sleeping here at all?

Makes the code a lot simpler, too, since you can await the receiver without the hardcoded `Ok(())` return value.

Reply:
Hmm I see what you are saying. My intent here was to "cleanly" exit the sink. Makes sense to make use of the full amount of time available though.

Do we emit an error in the logs if one of the components does not successfully shutdown within the grace period?
I would expect so, and then that is also better than what I have here. (Should have added a error log in the timeout case).

Reply:
Found this in running.rs:

```Failed to gracefully shut down in time. Killing components."```

Updating the code...

Reply:
I'm leaning towards adding these log messages:
  - (WARN) stating that we need to send [x,y,x] payloads before shutting down
  - (INFO) stating when we sent those payloads

This will provide us evidence in the logs that not all stats payloads were flushed and how many didn't make it, if that happens.
Thoughts on that?

Reply:
However :thinking: that should realistically never happen.
Because we are only ever keeping two payloads cached.
+ future.
The most then that we would need to flush is 3 payloads.

I think its highly unlikely that this thread will ever get killed because of not ending soon enough.
Unless there are issues performing the HTTP requests :thinking: 

Can't decide if that would just be annoying in the logs or would be helpful :joy:

Reply:
I don't think we need to log here.

As you've noted, Vector will already log when it failed to gracefully shutdown components. It will also log, on an interval, the components it's still waiting on to complete when in the shutdown grace period. Since the only thing it could be waiting on is the APM stats to be flushed, we can already deduce the answer to "why is the sink hanging?" from existing logging.